### PR TITLE
Implementation of Update Cart endpoint - Pre-step 2

### PR DESCRIPTION
**Pull request by dumega**
# Description
Add properties 'discount_category' and 'discount_type' to the discount item of Bolt cart, so the removal feature can be triggered properly

Prev PR https://github.com/BoltApp/bolt-magento2/pull/857
Next PR https://github.com/BoltApp/bolt-magento2/pull/852

Related task https://boltpay.atlassian.net/browse/M2P-64

Fixes: (link Jira ticket)

#changelog Implementation of Update Cart endpoint - Pre-step 2

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Helper/Cart.php [hunk 1]
```diff
@@ -1725,7 +1725,9 @@ public function collectDiscounts(
             $discounts[] = [
                 'description' => trim(__('Discount ') . $address->getDiscountDescription()),
                 'amount'      => $roundedAmount,
-                'reference'   => $address->getCouponCode()
+                'reference'   => $address->getCouponCode(),
+                'discount_category' => 'coupon',
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1729]: Could you please add constants for discount_category, magento discount types and bolt discount types?

Comment (dumega) [reply Helper/Cart.php]: Done

File: Helper/Cart.php [hunk 2]
```diff
@@ -1754,7 +1754,7 @@ public function buildCartFromQuote($quote, $immutableQuote, $items, $placeOrderP
         if ($totalAmount < 0) {
             $totalAmount = 0;
             foreach ($cart['discounts'] as &$discount) {
-                $discount['type'] = 'fixed_amount';
```
Comment (danac-gs) [inline Helper/Cart.php:1757]: This change does not seem (backward) compatible with the current flow. You should probably leave the `type` field as well. Or verify the impact and be 100% sure that `discount_type` and `type` are treated as synonyms on Bolt side. There's functionality around this that needs to stay - when the amount of discount / gift certificate / store credit / reward points is above the order total.

Comment (matt-thomason) [reply Helper/Cart.php:1757]: Good catch here @danac-gs and I agree. We can't guarantee all merchants who auto-update to this version are going to work with their CSM to properly configure the new endpoint.

Comment (dumega) [reply Helper/Cart.php:1757]: @danac-gs @matt-thomason I have discussed with @melvin0008 , and for now we do need both of the two properties, so `type` is for v1/merchant/order, `discount_type ` is for discount.apply and v2/update-cart

Then I have fixed this, please review again. Thanks

File: Helper/Cart.php [hunk 3]
```diff
@@ -1862,6 +1865,8 @@ public function collectDiscounts(
                 $discounts[] = [
                     'description' => 'Store Credit',
                     'amount'      => $roundedAmount,
+                    'discount_category' => Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT,
+                    'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
```
Comment (danac-gs) [inline Helper/Cart.php:1869]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 4]
```diff
@@ -1883,7 +1888,8 @@ public function collectDiscounts(
                     $discounts[] = [
                         'description' => 'Store Credit',
                         'amount'      => $roundedAmount,
-                        'type'        => 'fixed_amount',
+                        'discount_category' => Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT,
+                        'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
```
Comment (danac-gs) [inline Helper/Cart.php:1892]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 5]
```diff
@@ -1902,7 +1908,8 @@ public function collectDiscounts(
             $discounts[] = [
                 'description' => 'Store Credit',
                 'amount'      => $roundedAmount,
-                'type'        => 'fixed_amount',
+                'discount_category' => Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT,
+                'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
```
Comment (danac-gs) [inline Helper/Cart.php:1912]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 6]
```diff
@@ -1919,7 +1926,8 @@ public function collectDiscounts(
...
```
Comment (danac-gs) [inline Helper/Cart.php:1930]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 7]
```diff
@@ -1940,7 +1948,8 @@ public function collectDiscounts(
...
```
Comment (danac-gs) [inline Helper/Cart.php:1952]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 8]
```diff
@@ -1958,6 +1967,8 @@ public function collectDiscounts(
                 $discounts[] = [
                     'description' => 'Reward Points',
                     'amount'      => $roundedAmount,
+                    'discount_category' => Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT,
+                    'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
```
Comment (danac-gs) [inline Helper/Cart.php:1971]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 9]
```diff
@@ -1979,7 +1990,8 @@ public function collectDiscounts(
                     $discounts[] = [
                         'description' => 'Reward Points',
                         'amount'      => $roundedAmount,
-                        'type'        => 'fixed_amount',
+                        'discount_category' => Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT,
+                        'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
```
Comment (danac-gs) [inline Helper/Cart.php:1994]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 10]
```diff
@@ -2003,7 +2015,8 @@ public function collectDiscounts(
                         $quote->getStoreId()
                     ),
                 'amount'      => $roundedAmount,
-                'type'        => 'fixed_amount',
+                'discount_category' => Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT,
+                'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
```
Comment (danac-gs) [inline Helper/Cart.php:2019]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 11]
```diff
@@ -2057,9 +2070,18 @@ public function collectDiscounts(
                     'description' => $description . @$totals[$discount]->getTitle(),
                     'amount'      => $roundedAmount,
                 ];
-
+                
+                if ($discount == Discount::GIFT_CARD_ACCOUNT
+                    || $discount == Discount::AMASTY_GIFTCARD
+                    || $discount == Discount::MAGEPLAZA_GIFTCARD
+                    || $discount == Discount::UNIRGY_GIFT_CERT) {
+                    $discountItem['discount_type'] = $this->discountHelper->getBoltDiscountType('by_fixed');
```
Comment (danac-gs) [inline Helper/Cart.php:2078]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Helper/Cart.php [hunk 12]
```diff
@@ -2057,9 +2070,18 @@ public function collectDiscounts(
...
+                    $discountItem['discount_category'] = Discount::BOLT_DISCOUNT_CATEGORY_GIFTCARD;
+                }
+                
                 if ($discount == Discount::AMASTY_STORECREDIT) {
-                    $discountItem['type'] = 'fixed_amount';
+                    $discountItem['discount_type'] = $this->discountHelper->getBoltDiscountType('by_fixed');
+                    $discountItem['discount_category'] = Discount::BOLT_DISCOUNT_CATEGORY_STORE_CREDIT;
```
Comment (danac-gs) [inline Helper/Cart.php:2084]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 13]
```diff
@@ -2662,7 +2667,11 @@ public function getCartData_whenPaymentOnlyAndHasOrderPayload_returnsCartData()
         $testDiscounts = [
             [
                 'description' => 'Test discount',
-                'amount'      => 1000
+                'amount'      => 1000,
+                'reference'   => self::COUPON_CODE,
+                'discount_category' => 'coupon',
+                'discount_type'   => 'fixed_amount',
+                'type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:2674]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 14]
```diff
@@ -3665,7 +3683,16 @@ public function collectDiscounts_withCouponCode_collectsCouponCodeDiscount()
...
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => trim(__('Discount ') . self::COUPON_DESCRIPTION),
+                'amount'      => $expectedDiscountAmount,
+                'reference'   => self::COUPON_CODE,
+                'discount_category' => 'coupon',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:3692]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 15]
```diff
@@ -3711,7 +3739,15 @@ public function collectDiscounts_withStoreCreditAndPaymentOnly_collectsDiscountF
         static::assertEquals($diffResult, $diff);
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => 'Store Credit',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:3747]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 16]
```diff
@@ -3782,10 +3819,15 @@ public function collectDiscounts_withStoreCreditAndNotPaymentOnly_loadsCustomerB
...
-
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
-        static::assertEquals('fixed_amount', $discounts[0]['type']);
-        static::assertEquals('Store Credit', $discounts[0]['description']);
+        $expectedDiscount = [
+            [
+                'description' => 'Store Credit',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:3827]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 17]
```diff
@@ -3851,10 +3894,15 @@ public function collectDiscounts_withRewardPointsAndNotPaymentOnly_loadsRewardPo
...
-
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
-        static::assertEquals('fixed_amount', $discounts[0]['type']);
-        static::assertEquals('Reward Points', $discounts[0]['description']);
+        $expectedDiscount = [
+            [
+                'description' => 'Reward Points',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:3902]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 18]
```diff
@@ -3916,10 +3965,15 @@ public function collectDiscounts_withMirasvitRewardPoints_collectsDiscount()
...
-
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
-        static::assertEquals('fixed_amount', $discounts[0]['type']);
-        static::assertEquals('Mirasvit Reward Points', $discounts[0]['description']);
+        $expectedDiscount = [
+            [
+                'description' => 'Mirasvit Reward Points',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:3973]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 19]
```diff
@@ -3961,7 +4016,15 @@ public function collectDiscounts_withMirasvitStoreCredit_collectsDiscount()
         static::assertEquals($diffResult, $diff);
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => 'Store Credit',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4024]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 20]
```diff
@@ -4006,7 +4070,15 @@ public function collectDiscounts_withRewardPointsAndPaymentOnly_collectsRewardPo
         static::assertEquals($diffResult, $diff);
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => 'Reward Points',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4078]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 21]
```diff
@@ -4056,7 +4129,15 @@ public function collectDiscounts_withAheadworksStoreCredit_collectsAheadworksSto
         static::assertEquals($diffResult, $diff);
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => 'Store Credit',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4137]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 22]
```diff
@@ -4105,7 +4187,15 @@ public function collectDiscounts_withBssStoreCredit_collectsBssStoreCredit()
...
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4195]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 23]
```diff
@@ -4160,7 +4251,15 @@ public function collectDiscounts_withAmastyGiftcard_collectsAmastyGiftcard()
         static::assertEquals($diffResult, $diff);
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => 'Gift Card ',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'giftcard',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4259]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 24]
```diff
@@ -4209,9 +4309,17 @@ public function collectDiscounts_withAmastyStoreCreditInQuoteTotals_collectsAmas
             $paymentOnly,
             $quote
         );
+        $expectedDiscount = [
+            [
+                'description' => 'Store Credit',
+                'amount'      => $appliedDiscount * 100,
+                'discount_category' => 'store_credit',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4317]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 25]
```diff
@@ -4268,7 +4377,15 @@ public function collectDiscounts_withMageplazaGiftCard_collectsMageplazaGiftCard
         static::assertEquals($diffResult, $diff);
         $expectedDiscountAmount = 100 * $appliedDiscount;
         $expectedTotalAmount = $totalAmount - $expectedDiscountAmount;
-        static::assertEquals($expectedDiscountAmount, $discounts[0]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => '',
+                'amount'      => $expectedDiscountAmount,
+                'discount_category' => 'giftcard',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4385]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 26]
```diff
@@ -4315,7 +4433,15 @@ public function collectDiscounts_withUnirgyGiftcert_collectsUnirgyGiftcert()
...
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4441]: We might need this`$discount['type'] = 'fixed_amount'` here as well

File: Test/Unit/Helper/CartTest.php [hunk 27]
```diff
@@ -4363,8 +4491,20 @@ public function collectDiscounts_withGiftVoucher_collectsGiftVoucher()
...
         $expectedTotalAmount = $totalAmount - $expectedRegularDiscountAmount - $expectedGiftVoucherAmount;
-        static::assertEquals($expectedRegularDiscountAmount, $discounts[0]['amount']);
-        static::assertEquals($expectedGiftVoucherAmount, $discounts[1]['amount']);
+        $expectedDiscount = [
+            [
+                'description' => trim(__('Discount ') . self::COUPON_DESCRIPTION),
+                'amount'      => $expectedRegularDiscountAmount,
+                'reference'   => $giftVaucher,
+                'discount_category' => 'coupon',
+                'discount_type'   => 'fixed_amount',
```
Comment (danac-gs) [inline Test/Unit/Helper/CartTest.php:4500]: We might need this`$discount['type'] = 'fixed_amount'` here as well

**Comment by dumega** [comment]
@danac-gs I add the `type` property per your comment, to avoid any backward compatibility.

**Pull request merged**